prisma/schema.prisma

```diff
@@ -10,8 +10,145 @@ datasource db {
+  startDate  DateTime     @db.Timestamptz()
+  endDate    DateTime     @db.Timestamptz()
+  createdAt  DateTime     @default(now()) @db.Timestamptz()
+  VoyageTeam VoyageTeam[]
+}
+
+model VoyageStatus {
+  id               String             @id @default(uuid())
+  name             String             @unique
+  description      String
+  createdAt        DateTime           @default(now()) @db.Timestamptz()
+  VoyageTeam       VoyageTeam[]
+  VoyageTeamMember VoyageTeamMember[]
+}
+
+model Tier {
+  id          String       @id @default(uuid())
+  name        String       @unique
+  description String
+  createdAt   DateTime     @default(now()) @db.Timestamptz()
+  VoyageTeam  VoyageTeam[]
+}
+
+model VoyageTeam {
+  id            String        @id @default(uuid())
+  voyage        Voyage?       @relation(fields: [voyageId], references: [id], onUpdate: Cascade, onDelete: SetNull)
+  voyageId      String?
+  name          String        @unique
+  status        VoyageStatus? @relation(fields: [statusId], references: [id], onUpdate: Cascade, onDelete: SetNull)
+  statusId      String?
+  repoUrl       String
+  repoUrlBE     String?
+  deployedUrl   String?
+  deployedUrlBE String?
+  tier          Tier?         @relation(fields: [tierId], references: [id], onUpdate: Cascade, onDelete: SetNull)
+  tierId        String?
+  endDate       DateTime
+  createdAt     DateTime      @default(now()) @db.Timestamptz()
+
+  VoyageTeamMember  VoyageTeamMember[]
+  TeamTechStackItem TeamTechStackItem[]
+}
+
+model VoyageRole {
+  id               String             @id @default(uuid())
+  name             String             @unique
+  description      String
+  createdAt        DateTime           @default(now()) @db.Timestamptz()
+  VoyageTeamMember VoyageTeamMember[]
+}
+
+model VoyageTeamMember {
+  id                    String                  @id @default(uuid())
+  member                User                    @relation(fields: [userId], references: [id], onUpdate: Cascade, onDelete: Cascade)
+  userId                String
+  voyageTeam            VoyageTeam              @relation(fields: [voyageTeamId], references: [id], onUpdate: Cascade, onDelete: Cascade)
+  voyageTeamId          String
+  voyageRole            VoyageRole?             @relation(fields: [voyageRoleId], references: [id], onUpdate: Cascade, onDelete: SetNull)
+  voyageRoleId          String?
+  status                VoyageStatus?           @relation(fields: [statusId], references: [id], onUpdate: Cascade, onDelete: SetNull)
+  statusId              String?
+  hrPerSprint           Float
+  createdAt             DateTime                @default(now()) @db.Timestamptz()
+  TeamTechStackItemVote TeamTechStackItemVote[]
+  ProjectIdea           ProjectIdea[]
+}
+
+model ProjectIdea {
+  id            String            @id @default(uuid())
+  contributedBy VoyageTeamMember? @relation(fields: [userId], references: [id], onUpdate: Cascade, onDelete: SetNull)
+  userId        String?
```

Comment:
Shouldn't the project idea have always a user associated with?

Reply:
yes, as above, maybe the user is deleted (maybe they dropped voyage, but it should never happen) for some reason and we still want to keep the ideas

Reply:
> yes, as above, maybe the user is deleted (maybe they dropped voyage, but it should never happen) for some reason and we still want to keep the ideas

If we need to worry about user being deleted, I'd agree, but if not I think it would be good to still keep the user (even if they drop the voyage) associated with project idea because eventually we will be expanding this project and having a much larger scale of a portfolio or history, if you will, of the user and their contributions in the chingu community and their voyages.

Reply:
Even if the user drops, a contribution was still made, so we need to think how we will handle that

Reply:
> Even if the user drops, a contribution was still made, so we need to think how we will handle that

that's another issue, drop !== delete, if they are dropped they would just have a "dropped" status, as I said, user deletion should never happen. But we need it here, so I can add an onDelete referential action

The options are 
[Cascade](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions#cascade)
[Restrict](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions#restrict)
[NoAction](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions#noaction)
[SetNull](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions#setnull)
[SetDefault](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions#setdefault)

We don't want cascade as it will delete all the ideas the users have made
Or we can use "restrict/no action" but restricts deletion of user with ideas - not sure about it if we actually have to delete something
set default - don't think there can be a "Default user"
SetNull seems to be the most logical here to me, we can easily filter stuff with Null in that field if we have to

Reply:
In the VoyageUser table there is a status that can have a special name that will flag us that is a user that is not longer working in the project, that way in this table we will still have the contributedBy field and its value can be not nullable

Reply:
> In the VoyageUser table there is a status that can have a special name that will flag us that is a user that is not longer working in the project, that way in this table we will still have the contributedBy field and its value can be not nullable

That's not what we are talking about

"we need it here, so I can add an onDelete referential action", what other onDelete referential do you suggest? we're talking about "onDelete" referential, when/if for any reason the user is deleted. not when the user status is updated. Why do we need this? because migrate will throw an error if I don't

Reply:
Having that status may enable us to have this relation as a "onDelete = restrict"

Reply:
> Having that status may enable us to have this relation as a "onDelete = restrict"

then what if you actually have to delete it? 🤷

Reply:
I don't even know why is it such a big deal if it is null. I don't understand why you have to go into such an efford to not want it to be null. Values are allowed to be null in the database sometimes. You just check if it's NULL before using the value.

Reply:
In general, and with few exceptions, data shouldn't be deleted.  It is a very common practice to have people drop down from a voyage, not only the first sprint, but also in the later stages of the projects.  Having the user deleted will also "erase" all the contributions made by she/him and that is something I don't agree with.

All data inside the tables should have meaning, and being null has to have a meaning, in this case I don't see the meaning of having it s null, but the loss of the information on who suggested something

Reply:
In that case we should have a process that cleans up everything the user has, and even then I'm not sure if it is the proper way to handle it.

It might have escaped me the gender field, it should have a default value of "PREFER NOT TO SAY", for instance I think that table is a good example of where we could not use UUID as the id type, but if you want to still use it is fine

Reply:
Personally I think every id other than user id can be auto increment and not uuid.

Reply:
I think we needn't worry as much about it as we are. Allowing a field to be nullable does not mean we actually need to make it null, there are other ways of dealing with it when necessary. But if it's makes it easier to manage now, let's go with the course of least resistance. As we get closer to mvp it's always possible to make revisions if necessary.

Reply:
> It might have escaped me the gender field, it should have a default value of "PREFER NOT TO SAY",

techincally, "prefer not to say" is not the same as not selecting anything (don't care)

but I think we all have a majority now to leave it as nullable, for multiple reasons